File: backend/app/api/tasks.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from typing import Optional
from datetime import datetime
from app.database.database import get_db
from app.models.user import User
from app.models.scraping_task import ScrapingTask
from app.schemas.task import TaskOut, TaskListResponse, TaskListItem
from app.core.security import get_current_user

router = APIRouter()


@router.get("/tasks", response_model=TaskListResponse)
def list_tasks(
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    status: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    query = db.query(ScrapingTask).filter(ScrapingTask.user_id == current_user.id)
    if status:
        query = query.filter(ScrapingTask.status == status.upper())
    total = query.count()
    tasks = (
        query.order_by(ScrapingTask.created_at.desc())
        .offset((page - 1) * per_page)
        .limit(per_page)
        .all()
    )
    import math
    return TaskListResponse(
        tasks=[TaskListItem.model_validate(t) for t in tasks],
        total=total,
        page=page,
        per_page=per_page,
        pages=math.ceil(total / per_page) if total else 1,
    )


@router.get("/tasks/{task_id}", response_model=TaskOut)
def get_task(
    task_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = (
        db.query(ScrapingTask)
        .filter(ScrapingTask.task_id == task_id, ScrapingTask.user_id == current_user.id)
        .first()
    )
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.delete("/tasks/{task_id}", status_code=204)
def delete_task(
    task_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = (
        db.query(ScrapingTask)
        .filter(ScrapingTask.task_id == task_id, ScrapingTask.user_id == current_user.id)
        .first()
    )
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    db.delete(task)
    db.commit()


@router.post("/tasks/{task_id}/cancel")
def cancel_task(
    task_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = (
        db.query(ScrapingTask)
        .filter(ScrapingTask.task_id == task_id, ScrapingTask.user_id == current_user.id)
        .first()
    )
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    if task.status not in ("PENDING", "RUNNING"):
        raise HTTPException(status_code=400, detail="Task cannot be cancelled in current state")
    task.status = "CANCELLED"
    task.completed_at = datetime.utcnow()
    db.commit()
    return {"message": "Task cancelled", "task_id": task_id}

File: backend/src/routes/tasks.js
const express = require('express');
const router = express.Router();
const { ScrapingTask } = require('../models');
const { authenticateToken } = require('../middleware/auth');

router.use(authenticateToken);

// GET /api/tasks
router.get('/tasks', async (req, res, next) => {
  try {
    const page = Math.max(1, parseInt(req.query.page || '1', 10));
    const perPage = Math.min(100, Math.max(1, parseInt(req.query.per_page || '20', 10)));
    const status = req.query.status ? String(req.query.status).toUpperCase() : null;

    const where = { user_id: req.user.id };
    if (status) {
      where.status = status;
    }

    const { count, rows } = await ScrapingTask.findAndCountAll({
      where,
      order: [['created_at', 'DESC']],
      offset: (page - 1) * perPage,
      limit: perPage,
    });

    const tasks = rows.map(t => ({
      id: t.id,
      task_id: t.task_id,
      status: t.status,
      location: t.location,
      keyword: t.keyword,
      results_discovered: t.results_discovered,
      websites_found: t.websites_found,
      created_at: t.created_at,
      completed_at: t.completed_at,
    }));

    return res.json({
      tasks,
      total: count,
      page,
      per_page: perPage,
      pages: count > 0 ? Math.ceil(count / perPage) : 1,
    });
  } catch (err) {
    next(err);
  }
});

// GET /api/tasks/:taskId
router.get('/tasks/:taskId', async (req, res, next) => {
  try {
    const { taskId } = req.params;
    const task = await ScrapingTask.findOne({
      where: {
        task_id: taskId,
        user_id: req.user.id,
      },
    });

    if (!task) {
      return res.status(404).json({ detail: 'Task not found' });
    }

    return res.json({
      id: task.id,
      task_id: task.task_id,
      status: task.status,
      location: task.location,
      keyword: task.keyword,
      search_radius: task.search_radius,
      max_results: task.max_results,
      max_pages_per_site: task.max_pages_per_site,
      results_discovered: task.results_discovered,
      websites_found: task.websites_found,
      websites_crawled: task.websites_crawled,
      phones_found: task.phones_found,
      emails_found: task.emails_found,
      addresses_found: task.addresses_found,
      duplicates_removed: task.duplicates_removed,
      error_message: task.error_message,
      created_at: task.created_at,
      updated_at: task.updated_at,
      completed_at: task.completed_at,
    });
  } catch (err) {
    next(err);
  }
});

// DELETE /api/tasks/:taskId
router.delete('/tasks/:taskId', async (req, res, next) => {
  try {
    const { taskId } = req.params;
    const task = await ScrapingTask.findOne({
      where: {
        task_id: taskId,
        user_id: req.user.id,
      },
    });

    if (!task) {
      return res.status(404).json({ detail: 'Task not found' });
    }

    const {
      Organization,
      Website,
      Contact,
      PhoneNumber,
      EmailAddress,
      SourcePage,
      SocialLink,
      SavedLead,
      ScrapingLog,
      Export,
    } = require('../models');

    const orgs = await Organization.findAll({
      where: { task_id: task.id },
      attributes: ['id'],
    });
    const orgIds = orgs.map(o => o.id);

    if (orgIds.length > 0) {
      const { Op } = require('sequelize');
      await SavedLead.destroy({ where: { organization_id: { [Op.in]: orgIds } } });
      await Website.destroy({ where: { organization_id: { [Op.in]: orgIds } } });
      await Contact.destroy({ where: { organization_id: { [Op.in]: orgIds } } });
      await PhoneNumber.destroy({ where: { organization_id: { [Op.in]: orgIds } } });
      await EmailAddress.destroy({ where: { organization_id: { [Op.in]: orgIds } } });
      await SourcePage.destroy({ where: { organization_id: { [Op.in]: orgIds } } });
      await SocialLink.destroy({ where: { organization_id: { [Op.in]: orgIds } } });
      await Organization.destroy({ where: { id: { [Op.in]: orgIds } } });
    }

    await ScrapingLog.destroy({ where: { task_id: task.id } });
    await Export.destroy({ where: { task_id: task.id } });
    await task.destroy();

    return res.status(204).send();
  } catch (err) {
    next(err);
  }
});

// POST /api/tasks/:taskId/cancel
router.post('/tasks/:taskId/cancel', async (req, res, next) => {
  try {
    const { taskId } = req.params;
    const task = await ScrapingTask.findOne({
      where: {
        task_id: taskId,
        user_id: req.user.id,
      },
    });

    if (!task) {
      return res.status(404).json({ detail: 'Task not found' });
    }

    if (!['PENDING', 'RUNNING'].includes(task.status)) {
      return res.status(400).json({ detail: 'Task cannot be cancelled in current state' });
    }

    task.status = 'CANCELLED';
    task.completed_at = new Date();
    await task.save();

    return res.json({ message: 'Task cancelled', task_id: taskId });
  } catch (err) {
    next(err);
  }
});

module.exports = router;
